s3/src/main/scala/geotrellis/store/s3/util/S3RangeReader.scala

```diff
@@ -36,22 +36,29 @@ import java.net.URI
  */
 class S3RangeReader(
   request: => GetObjectRequest,
-  client: S3Client
+  client: S3Client,
+  readHeader: Boolean = true
 ) extends RangeReader {
 
-  lazy val metadata: HeadObjectResponse = {
-    val headRequest =
-      HeadObjectRequest
-        .builder()
-        .bucket(request.bucket)
-        .key(request.key)
-        .build()
-
-    client.headObject(headRequest)
+  lazy val totalLength: Long = {
+    if (readHeader) {
+      val headRequest =
+        HeadObjectRequest
+          .builder()
+          .bucket(request.bucket)
+          .key(request.key)
+          .build()
+
+      client.headObject(headRequest).contentLength
+    } else {
+      val responseStream = client.getObject(request)
```

Comment:
This is a neat trick if it works. I would say given how much we use this class a test is required. 

Unit test is not necessary but loading this branch in console and pointing at the largest GeoTiff you can find with `S3RangeReader` would be enough. If both header and this non-emptied InputStream provide the content length with-in the same time this is a nice win.

Reply:
@echeipesh Will do :+1: Is there a particular GeoTiff you'd recommend?

Reply:
@jbouffard I think you can upload any tiff from the test files an upload them into our bucket and check that you can read it with limited bucket query permissions.

Reply:
@echeipesh @pomadchin I did some reading from the console and it didn't _seem_ like there was a big difference in time between the `HeadRequest` and the `ResponseStream`, but I could easily be mistaken.

Reply:
@jbouffard ah, the issue was in permissions. Usually users in RF give RF user only ability to use Get requests / GetList and nothing more. The idea of this PR was to ensure that we can work even in the case we can't perform `HEAD` requests. Can you double-check that the S3 GetRequest pulls down data from a bucket with permissions only to `Get` file and that it can't to the same with a `Head` request?

Reply:
@jbouffard can you also double check it with this tiff? `s3://geotrellis-test/issue-1691/firstPart.tif`? It's a `3.1 GB` tiff.

Reply:
@echeipesh @pomadchin Unfortunately, it appears that using the `ResponseStream` will cause the client to pull down the whole file. I was able to confirm this using WireShark when finding the `totalLength` of `s3://geotrellis-test/issue-1691/firstPart.tif`.

Reply:
@echeipesh After our conversation yesterday, I was curious to see what would happen if you tried to read past the file length on S3; and it turns out that it does throw an exception:

```
scala> r2.readRange(r2.totalLength, 10)
software.amazon.awssdk.services.s3.model.S3Exception: The requested range is not satisfiable (Service: S3, Status Code: 416, Request ID: B073A86E409076E4)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.handleErrorResponse(HandleResponseStage.java:115)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.handleResponse(HandleResponseStage.java:73)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.execute(HandleResponseStage.java:58)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.execute(HandleResponseStage.java:41)
  at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:64)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:36)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:77)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:39)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage$RetryExecutor.doExecute(RetryableStage.java:113)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage$RetryExecutor.execute(RetryableStage.java:86)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:62)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:42)
  at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
  at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:57)
  at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:37)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.executeWithTimer(ApiCallTimeoutTrackingStage.java:80)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:60)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:42)
  at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
  at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:37)
  at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:26)
  at software.amazon.awssdk.core.internal.http.AmazonSyncHttpClient$RequestExecutionBuilderImpl.execute(AmazonSyncHttpClient.java:240)
  at software.amazon.awssdk.core.client.handler.BaseSyncClientHandler.invoke(BaseSyncClientHandler.java:96)
  at software.amazon.awssdk.core.client.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:120)
  at software.amazon.awssdk.core.client.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:61)
  at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:51)
  at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:62)
  at software.amazon.awssdk.services.s3.DefaultS3Client.getObject(DefaultS3Client.java:1597)
  at software.amazon.awssdk.services.s3.S3Client.getObject(S3Client.java:2363)
  at geotrellis.store.s3.util.S3RangeReader.readClippedRange(S3RangeReader.scala:72)
  at geotrellis.util.RangeReader$class.readRange(RangeReader.scala:41)
  at geotrellis.store.s3.util.S3RangeReader.readRange(S3RangeReader.scala:38)
  ... 40 elided
```